### ISSUE TITLE
ITypeConverter for custom generics (type constructors)

### ISSUE THREAD

**Issue by Krever** [remkop/picocli#2156]
Hey! I'm trying to use picocli with Scala and one of the main limitations is inability to define converters from scala stdlib. While simple types are ok, the problem arises with generic types (called type constructors in scala). 

Scala has its own `scala.List`, `scala.Map` and `scala.Option`. HAving to use java variants is a bit suboptimal. 

This might be related to https://github.com/remkop/picocli/issues/1804 but Im not sure if interface described there would suffice, so I'm raising a separate ticket.

**Comment by remkop** [comment]
Is the problem that you want to use Scala collection types for multi-value options and positional parameters?
So, in your program you want to annotate fields of type `scala.List`, `scala.Map` and `scala.Option` with picocli `@Option` and `@Parameters`?

Picocli internally uses reflection to see if multi-value fields implement the `java.util.Collection` interface or the `java.util.Map` interface, and works with these Java interfaces in the parser. Similar with `java.util.Optional`. This is fairly deeply embedded at the moment and would require some refactoring (and perhaps some additional API) to decouple.

However, I have one idea that should allow you to use picocli-annotated `scala.List` and `scala.Map` fields/methods in your Scala program with the current version of picocli: 
use a `@Option`-annotated method, whose parameter type is  `java.util.Collection` or `java.util.Map`, and in the implementation of that method, delegate to fields that use the Scala types.
Then, in the business logic of your application, you can just use the Scala-typed fields.

For example (PSEUDO CODE):

```
@Command
class App {

    scala.List myList;

    scala.Map myMap;

    @Option(names = Array("-f", "--file"), description = Array("The files to process. -fFile1 -fFile2"))
    def setFiles(files: java.util.List) {
        myList.clear()
        myList.addAll(files)
    }

    @Option(names = Array("-D", "--property"), description = Array("The properties. -Dkey1=val1 -Dkey2=val2"))
    def setProperties(properties: java.util.Map) {
        myMap.clear()
        myMap.putAll(properties)
    }
````

Can you give this a try?

**Comment by remkop** [comment]
@Krever Glad you found an efficient workaround.

Are you okay if I close this ticket? To be honest, I don't see myself working on API to support non-java Collection and Map-like data structures.

**Comment by Krever** [comment]
Hey @remkop, I understand, it's fine to close the issue. I think it might significantly limit the adoption from Scala (having working native collections is rather important). However, at the same time, I see value in focus (on Java/Kotlin) and understand the lack of resources to add this significant change. 

Thanks a lot for the responses :)

**Comment by bherw** [comment]
Thanks to @remkop for the workaround; I was able to use a similar approach to populate a Guava Multimap:

```java
Multimap<UUID, String> examples = MultimapBuilder.linkedHashKeys().linkedHashSetValues().build();

@Option(
		names = "--examples",
		paramLabel = "<uuid>=<string>"
)
public void setExamples(Map<UUID, String> newValues) {
	for (Map.Entry<UUID, String> entry : newValues.entrySet()) {
		examples.put(entry.getKey(), entry.getValue());
	}
}
```

Unlike the suggestion given above, this code doesn't ever clear the multimap, but this seems to work fine if PicoCLI doesn't need to remove anything--parsing arguments should just be additive.

**Comment by Krever** [comment]
Hey @remkop, I got back to this problem and thought of an alternative workaround. 

What do you think about traversing CommandSpec and injecting custom `IParameterConsumer` whenever a scala type is present (e.g. collection or `Option`)? 

The problem I currently have with this approach:
* I have `CommandLine`  in my `main`, from which I can get `CommandSpec`. 
* The `CommandSpec` seems mutable, so I can substitute `ArgSpec` with one with a custom consumer
* `ArgSpec` is not mutable (at least not its `parameterConsumer` field is not), so I should probably go through `ArgSpec.Builder`
* I don't see a way to go from `ArgSpec` to `ArgSpec.Builder` 

I would appreciate any hint or info that this approach won't fly for some reason.

**Comment by remkop** [comment]
I think there's an OptionSpec.Builder constructor that takes an existing OptionSpec, and similarly for PositionalParamSpec.Builder (constructor that takes a PositionalParamSpec). 
Is that what you're looking for?

**Comment by Krever** [comment]
Yes, that's it. Thank you! I haven't looked carefully enough, sorry!
